partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/PlanTyper.kt

```diff
@@ -223,16 +224,47 @@ internal class PlanTyper(private val env: Env) {
             return rel(type, op)
         }
 
-        override fun visitRelOpUnion(node: Rel.Op.Union, ctx: Rel.Type?): Rel {
-            TODO("Type RelOp Union")
-        }
-
-        override fun visitRelOpIntersect(node: Rel.Op.Intersect, ctx: Rel.Type?): Rel {
-            TODO("Type RelOp Intersect")
-        }
+        // TODO: [RFC-0007](https://github.com/partiql/partiql-lang/blob/main/RFCs/0007-rfc-bag-operators.md)
+        //  states that the types must be "comparable". The below code ONLY makes sure that types need to be
+        //  the same. In the future, we need to add support for checking comparable types.
+        override fun visitRelOpSet(node: Rel.Op.Set, ctx: Rel.Type?): Rel {
+            val lhs = visitRel(node.lhs, node.lhs.type)
+            val rhs = visitRel(node.rhs, node.rhs.type)
+            val set = node.copy(lhs = lhs, rhs = rhs)
+
+            // Check that types are comparable
+            if (!node.isOuter) {
+                if (lhs.type.schema.size != rhs.type.schema.size) {
+                    return Rel(Rel.Type(emptyList(), emptySet()), Rel.Op.Err("LHS and RHS of SET OP do not have the same number of bindings."))
+                }
+                for (i in 0..lhs.type.schema.lastIndex) {
+                    val lhsBindingType = lhs.type.schema[i].type
+                    val rhsBindingType = rhs.type.schema[i].type
+                    if (lhsBindingType != rhsBindingType) {
+                        return Rel(Rel.Type(emptyList(), emptySet()), Rel.Op.Err("LHS and RHS of SET OP do not have the same type."))
+                    }
+                }
+            }
 
-        override fun visitRelOpExcept(node: Rel.Op.Except, ctx: Rel.Type?): Rel {
-            TODO("Type RelOp Except")
+            // Compute Output Schema
+            val size = max(lhs.type.schema.size, rhs.type.schema.size)
```

Comment:
Would this not depend on both the OUTER boolean and the operator type? 

Like OUTER UNION size may not be computable with open tuples.